corehq/apps/es/README.rst

```diff
@@ -0,0 +1,220 @@
+  hoster are not outside the Dimagi main track.
+
+
+.. TODO: Future adapter documentation
+..
+.. - adapter patterns
+.. - querying with adapters
+.. - writing tests (use ``es_test``)
+
+----
+
+Elastic Client Adapters
+-----------------------
+
+The ``corehq.apps.es.client`` module encapsulates the CommCare HQ Elasticsearch
+client adapters. It implements a high-level Elasticsearch client protocol
+necessary to accomplish all interactions with the backend Elasticsearch cluster.
+Client adapters are split into two usage patterns, the "Management Adapter" and
+"Document Adapters".
+
+.. toctree::
+
+    Management Adapter
+    Document Adapters
+    Code Documentation
+
+
+Management Adapter
+''''''''''''''''''
+
+There is only one management adapter, ``ElasticManageAdapter``. This adapter is
+used for performing all cluster management tasks such as creating and updating
+indices and their mappings, changing index settings, changing cluster settings,
+etc.  This functionality is split into a separate class for a few reasons:
+
+1. The management adapter is responsible for low-level Elastic operations which
+   document adapters should never be performing because the scope of a document
+   adapter does not extend beyond a single index.
+2. Elasticsearch 5+ implements security features which limit the kinds of
+   operations a connection can be used for. The separation in these client
+   adapter classes is designed to fit into that model.
+
+The management adapter does not need any special parameters to work with, and
+can be instantiated and used directly:
+
+.. code-block:: python
+
+    adapter = ElasticManageAdapter()
+    adapter.index_create("books")
+    mapping = {"properties": {
+        "author": {"type": "text"},
+        "title": {"type": "text"},
+        "published": {"type": "date"},
+    }}
+    adapter.index_put_mapping("books", "book", mapping)
+    adapter.index_refresh("books")
+    adapter.index_delete("books")
+
+
+Document Adapters
+'''''''''''''''''
+
+Document adapters are created on a per-index basis and include specific
+properties and functionality necessary for maintaining a single type of "model"
+document in a single index.  Each index in Elasticsearch needs to have a
+cooresponding ``ElasticDocumentAdapter`` subclass which defines how the Python
+model is applied to that specific index.  At the very least, a document adapter
+must define the following:
+
+- An ``_index_name`` attribute whose value is the name of the Elastic index
+  used by the adapter. This attribute must be private to support proper index
+  naming between production code and tests.
+- A ``type`` attribute whose value is the name is the Elastic ``_type`` for
+  documents used by the adapter.
+- A ``mapping`` which defines the structure and properties for documents managed
+  by the adapter.
+- A ``from_python()`` classmethod which can convert a Python model object into the
```

Comment:
Second question first

> is the long term intention to move the full transform from the pillow processor to these adapter classes?

Yes.

> It looks like from_python takes a dictionary rather than a python model class. Is that correct?

By this are you talking about how all the current adapters (e.g. [ElasticUser.from_python()](https://github.com/dimagi/commcare-hq/blob/cc8c4d31e85/corehq/apps/es/users.py#L78-L80)) accept docs as dicts?  Those `from_python()` methods currently accept dicts because at this transient stage of the adapter implementation, pillows still owns and performs the transformation logic from python model instances to dicts. In order to avoid ripping that out of pillows and making this PR even larger, yes, the current _functional_ `from_python()` classmethods accept dicts.  They will not in the future.

Reply:
All of our current transformation functions from pillows operate on dictionaries rather than classes, does this mean the intention is to change that model?

Reply:
Which transformation functions? Functions that transform a document to be indexed _into_ Elasticsearch, or functions that transform a document fetched _from_ Elasticsearch?